include/swift/Basic/Statistic.h

```diff
@@ -110,20 +110,31 @@ class UnifiedStatsReporter {
   // active at all. Reduces redundant machinery.
   class RecursionSafeTimers;
 
+  // We also keep a few banks of optional hierarchical profilers for times and
+  // statistics, activated with -profile-stats-events and
+  // -profile-stats-entities, which are part way between the detail level of the
+  // aggregate statistic JSON files and the fine-grained CSV traces. Naturally
+  // these are written in yet a different file format: the input format for
+  // flamegraphs.
+  struct StatsProfilers;
+
 private:
   bool currentProcessExitStatusSet;
   int currentProcessExitStatus;
   SmallString<128> StatsFilename;
   SmallString<128> TraceFilename;
+  SmallString<128> ProfileDirname;
   llvm::TimeRecord StartedTime;
   std::unique_ptr<llvm::NamedRegionTimer> Timer;
   SourceManager *SourceMgr;
   clang::SourceManager *ClangSourceMgr;
   std::unique_ptr<AlwaysOnDriverCounters> DriverCounters;
   std::unique_ptr<AlwaysOnFrontendCounters> FrontendCounters;
   std::unique_ptr<AlwaysOnFrontendCounters> LastTracedFrontendCounters;
-  std::vector<FrontendStatsEvent> FrontendStatsEvents;
+  std::unique_ptr<std::vector<FrontendStatsEvent>> FrontendStatsEvents;
```

Comment:
Why the extra indirection?

Reply:
Congruent with the other machinery where we test if it's active based on the presence/absence of the supporting data structure (i.e. nullptr == disabled); this PR changes the test for tracing (vs. profiling which also uses the LastTracedFrontendCounters) from testing LastTracedFrontendCounters to testing FrontendStatsEvents, so the latter moves to a thing-that-can-be-nullptr. I suppose I could also split enabled/disabled signalling into a set of boolean flags if you prefer. That would make it a bit more obvious.

Reply:
I'm fine with an explicit test rather than relying on an empty state, but is there a reason to use `unique_ptr` over `Optional`?

Reply:
Nope, just muscle memory from the days before a reliable C++ Optional in my codebase.

Optional is good! I'll switch 'em over.